### PR TITLE
Refactorings from cluster trx improvement branch

### PR DESCRIPTION
**Pull request by graetzer**
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr-linux/4741/

### CODE REVIEW

File: arangod/RestHandler/RestReplicationHandler.cpp [hunk 1]
```diff
@@ -1007,17 +1007,18 @@ Result RestReplicationHandler::processRestoreCollectionCoordinator(
...
-
-        if (res != TRI_ERROR_NO_ERROR) {
-          return Result(
-              res,
-              std::string(
-                  "unable to truncate collection (dropping is forbidden): '") +
-                  name + "'");
+        auto ctx = transaction::StandaloneContext::Create(_vocbase);
+        SingleCollectionTransaction trx(ctx, name, AccessMode::Type::EXCLUSIVE);
+        trx.addHint(transaction::Hints::Hint::SINGLE_OPERATION); // immediately commit on follower
```
Comment (jsteemann) [inline arangod/RestHandler/RestReplicationHandler.cpp]: This does not look right here, as truncate will mostly likely contain multiple operations.

File: arangod/Transaction/Manager.cpp [hunk 2]
```diff
@@ -21,70 +21,84 @@
...
+    const size_t bucket = getBucket(transactionId);
+    READ_LOCKER(allTransactionsLocker, _allTransactionsLock);
 
-  WRITE_LOCKER(writeLocker, _transactions[bucket]._lock);
+    WRITE_LOCKER(writeLocker, _transactions[bucket]._lock);
 
-  // insert into currently running list of transactions
-  _transactions[bucket]._activeTransactions.emplace(transactionId, std::move(data));
+    // insert into currently running list of transactions
+    _transactions[bucket]._activeTransactions.emplace(transactionId, std::move(data));
```
Comment (jsteemann) [inline arangod/Transaction/Manager.cpp]: if this throws, is `_nrRunning` decremented again, or will it be off forever?

File: arangod/Transaction/Manager.h [hunk 3]
```diff
@@ -21,66 +21,87 @@
...
-class MMFilesTransactionManager final : public TransactionManager {
+class Manager final {
   static constexpr size_t numBuckets = 16;
+  static constexpr double defaultTTL = 10.0 * 60.0;   // 10 minutes
+  static constexpr double tombstoneTTL = 5.0 * 60.0;  // 5 minutes
 
  public:
-  MMFilesTransactionManager() : TransactionManager() {}
-  ~MMFilesTransactionManager() {}
+  Manager(bool keepData) : _keepTransactionData(keepData), _nrRunning(0) {}
```
Comment (jsteemann) [inline arangod/Transaction/Manager.h]: let's add `explicit` here.
```suggestion
  explicit Manager(bool keepData) : _keepTransactionData(keepData), _nrRunning(0) {}
```

File: arangod/Transaction/ManagerFeature.cpp [hunk 4]
```diff
@@ -20,35 +20,45 @@
...
+void ManagerFeature::prepare() {
   TRI_ASSERT(MANAGER == nullptr);
   TRI_ASSERT(EngineSelectorFeature::ENGINE != nullptr);
   MANAGER = EngineSelectorFeature::ENGINE->createTransactionManager();
 }
 
-void TransactionManagerFeature::unprepare() { MANAGER.reset(); }
+void ManagerFeature::start() {
+  auto off = std::chrono::seconds(1);
+  _workItem = SchedulerFeature::SCHEDULER->queueDelay(RequestLane::INTERNAL_LOW, off, _gcfunc);
```
Comment (jsteemann) [inline arangod/Transaction/ManagerFeature.cpp]: is `_gcfunc` ever set by any code? I think currently it is not.

**Comment by graetzer** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr-linux/4710/

File: arangod/Aql/Functions.cpp [hunk 5]
```diff
@@ -1466,6 +1466,7 @@ AqlValue Functions::LevenshteinDistance(ExpressionContext*, transaction::Methods
   AqlValue const& value1 = extractFunctionParameterValue(parameters, 0);
   AqlValue const& value2 = extractFunctionParameterValue(parameters, 1);
 
+  // FIXME: there is only one shared stringbuffer instance
```
[the remark this answers is not in the document]
Comment (graetzer) [reply arangod/Aql/Functions.cpp]: There is only one shared buffer, the second one is always allocated.
Anyway the goal is to delete the old StringBuffer an replace it with a shared std::string (vector)

**Comment by jsteemann** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/3428/

**Pull request merged**